### ISSUE TITLE
[Request] Documentation: add examples of named list elements in `i`

### ISSUE THREAD

**Issue by Henrik-P** [Rdatatable/data.table#1945]
As is also stated in the documentation, the `on` argument indeed makes code for subset / join easier to understand. However, the values used for subset/join provided in `i` are most often written as an unnamed `i` argument at the leftmost part of `[ ]`, while their corresponding variable names in `on` are often written 'far away' to the right in `[ ]`. E.g.: 

    flights[.(c("LGA", "JFK", "EWR"), c("LAX", "PBI", "MIA")), j = verylongcalculation, keyby = somelongvariablenames, mult = "last", nomatch = NA, on = c("origin", "dest")]

Thus, it may be slightly awkward to link the variable names in `on` and their corresponding values in `i`. On the other hand, having named elements in the `i` list explicitly link the values of the objects to their corresponding variable in `on` (or in the `key`).

It may well be obvious for many (most) `data.table` users that list elements used in `i` may be named. However, I believe that by adding some examples on such 'name-value' pairs in `i`, it would hopefully support code which is easier to follow. 

Here's a suggestion of some cases in the documentation which you may consider to supplement with named list examples as well.

In `?data.table`: 

```
# all together now
DT[c("b","c"), sum(v), by=.EACHI, on=.(x)]
DT[.(x = c("b", "c")), sum(v), by = .EACHI, on = .(x)] # same, but more explicit by using named list elements in i


# fast *keyed* subsets on multi-column key
kDT[.("a", 3)]
kDT[.(x = "a", y = 3)] # same, but more explicit by using named list elements in i

```

In the vignette [Secondary indices and auto indexing](https://rawgit.com/wiki/Rdatatable/data.table/vignettes/datatable-secondary-indices-and-auto-indexing.html):

```
flights[.("LGA", "TPA"), max(arr_delay), on = c("origin", "dest")]

# same, but more explicit by using named list elements in i
flights[.(origin = "LGA", dest = "TPA"), max(arr_delay), on = c("origin", "dest")]


flights[.(c("LGA", "JFK", "EWR"), "XNA"), mult = "last", on = c("origin", "dest"), nomatch = 0L]

# same, but more explicit by using named list elements in i
flights[.(origin = c("LGA", "JFK", "EWR"), dest = "XNA"), mult = "last", on = c("origin", "dest"), nomatch = 0L]
```

**Comment by Henrik-P** [comment]
It may be worth noting though that the order of explicit variable-value pairs in a named list in `i` is respected only when `on` is used. Without `on`, the order of keys can not be overruled by the order of the named list in `i`.

    set.seed(1)
    d <- data.table(x = sample(letters[1:3], 10, replace = TRUE),
                    y = sample(letters[1:3], 10, replace = TRUE),
                    z = 1:10)
    setkey(d, x, y)

When `on` is used, the named list "works", also when the order of variables in `i` differs from the order in `on`:
 
    d[.(y = "b", x = "c"), on = .(x, y)]
    #    x y z
    # 1: c b 4
    # 2: c b 6

However, without `on`, `y =` and `x =` have no effect - `"b"` and `"c"` and is matched to variables in `d` only according to the order of the keys of `d` (`x, y`). 
 
    d[.(y = "b", x = "c")]
    #    x y z
    # 1: b c 3
    # 2: b c 8

**Comment by franknarf1** [comment]
Some minor comments:

> the values used for subset/join provided in i are most often written as an unnamed i argument at the leftmost part of [ ], while their corresponding variable names in on are often written 'far away' to the right in [ ]. 

That's at the user's discretion. Personally, I'll typically do

```
x[i, on=, roll=, mult=, nomatch=, j]
```

This ordering shows up in one example in the docs at least:

```
DT[X, on=.(x, v>=v), sum(y)*foo, by=.EACHI] # NEW non-equi join with by=.EACHI (v1.9.8+)
```

----

Regarding stuff in `i` being named, that's thanks to `i` being allowed to be a data.table. Similarly, the part of the documentation that covers when `i` is a data.table explains your second comment about joining on column order when `x` is keyed but `i` is not.

----

Most of this would fit most naturally in the joins vignette ( #944 ) rather than the secondary indices one or `example(data.table)`, I guess.

**Comment by MichaelChirico** [comment]
@franknarf1 I agree it's basically covered/implied by the dox, but any time an experienced user scratches their head at a behavior, it's probably worth adding a quick reference in the main documentation (with a mind for not obfuscating any current info, of course).

**Comment by venom1204** [comment]
Hi @MichaelChirico @jangorecki 

I hope this message finds you well. I came across issue #1945 regarding adding examples of named list elements in i to the data.table documentation. I wanted to check if this issue is still relevant and if so, I would love to work on it.

My plan is to update the documentation in the following files:

vignettes/datatable-intro.Rmd

vignettes/datatable-secondary-indices-and-auto-indexing.Rmd

vignettes/datatable-joins.Rmd

man/data.table.Rd

I intend to add clear examples that demonstrate how using named list elements in i can enhance code readability and maintainability. This will include both simple and more complex scenarios to cover a wide range of use cases.

Before proceeding, I'd appreciate your confirmation on whether this issue is still open and if my approach aligns with your expectations.

Thank you for your time, and I look forward to hearing back from you.

**Comment by jangorecki** [comment]
While I agree some doc could mention named list in i, I do not think it is good idea to change it everywhere as it is overly verbose which is against data.table concise syntax.
